contracts/IntervalTokenStore.sol

```diff
@@ -0,0 +1,142 @@
+// IntervalTokenStore stores tokens for applications, which have discrete states increasing with time
+pragma solidity ^0.5.0;
+
+import "openzeppelin-solidity/contracts/token/ERC20/ERC20.sol";
+import "openzeppelin-solidity/contracts/math/Math.sol";
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+
+contract IntervalTokenStore {
+    using SafeMath for uint;
+
+    event Deposit(
+        address user,
+        address token,
+        uint amount,
+        uint stateIndex
+    );
+
+    event WithdrawRequest(
+        address user,
+        address token,
+        uint amount,
+        uint stateIndex
+    );
+
+    event Withdraw(
+        address user,
+        address token,
+        uint amount
+    );
+
+    // User => Token => BalanceState
+    mapping(address => mapping(address => BalanceState)) private balanceStates;
+
+    struct BalanceState {
+        uint256 balance;
+        PendingFlux pendingDeposits;
+        PendingFlux pendingWithdraws;
+    }
+
+    struct PendingFlux {
+        uint256 amount;
+        uint256 stateIndex;  // deposits will be processed for any currentStateIndex > stateIndex
+                              // withdraws are allowed for any currentStateIndex > stateIndex
+    }
+
+    uint256 public currentStateIndex=0;
+
+    function updateAndGetBalance(address user, address token) public returns(uint256) {
+        updateDepositsBalance(user, token);
```

Comment:
What's the need to update the balance here? Can't we just update whenever we deposit and leave this function view?

Reply:
Later on, when we trade, we want to make sure that we are always working with the latest balance available. Hence, we enforce any reading through the interfaces:
`updateAndGetBalance, addBalance, substractBalance` and all of them make sure that the balances are updated before any other operation will happen.

This is my current thinking, maybe, you are aware of  a cleaner way to do it. :)

Reply:
I agree that all writes to the balance should update pending deposits (and there should be an interface that people are forced to use). However, I feel like we should allow people to read the balances without requiring a tx (e.g. the driver will want to query all balances before it starts the solver without sending n transactions).

Reply:
I don't have super strong opinion here, but I think you don't need a tx for getting the balance.

On the one hand, there's already this view function that shows the unupdated balance:
```
function getBalance(address user, address token) public view returns(uint) {
        return balanceStates[user][token].balance;
    }
```

And in the other, if you do a `call` instead of a `tx`, you would get the "real balance" taking the pending amounts into account, but it doesn't modify the blockchain or cost anything the transactions.

I also see it a bit strange a method that is called `updateAndGetBalance`, @fleupold does it sound better if it's just renamed to `updateBalance` (it's normal that a updateBalance returns the value of the updated balance).

Reply:
Yes, I think that name would be better. I'd still argue there should be a view function just returning the current balance (without writing it).

Reply:
Hm...
I feel that EpochTokenLocker should have a defined interface with a unique method to `get` the current balance. And due to this unusual design pattern, we need to update and read the balance at the same time. That is also why I would prefer to have both getBalance and update in the name. 
If we have the only `updateBalance`, people might wonder where they can get the balance from in the first place. I think it would be less intuitive.

 >I'd still argue there should be a view function just returning the current balance (without writing it).

I am happy to add another function. This could be useful for other smart contracts, which want to act conditionally on the exchange and don't have the ability to do a .call. Is this the purpose you have in mind, Felix?

Reply:
> And due to this unusual design pattern, we need to update and read the balance at the same time.

We don't really have to update the balance whenever we want to read it. We could just evaluate the pending deposit and leave it there. We only really need to update when we want to mutate the balance, which is already a tx (I think your suggestion makes sense to only allow that via an interface we control).

However, I think that the reading and writing interface should be separate. Therefore I advertise for a view function that allows to read the balance of the current user that doesn't have side effects.

Reply:
I think it's similar as banks, when you receive a transfer and is not settled, the bank shows both the total balance and the available balance (at least my bank :P )

I find it ok to show both, it makes sense to me too Felix suggestion of having two view fn:
* `getTotalBalance`
* `getAvailableBalance`

Additionally we would need the `updateBalance` or `applyPendingBalance`

Reply:
Felix and I decided that actually only the getTotalBalance is needed. Other value is not of relevance.

and I implemented the updateBalance for sure

Reply:
But why? 

I still think it's relevant something like:
* *Total balance*: balance + applicable deposits - appliable withdraws
* *Available balance*: balance + applicable deposits - *any withdraw*

Also, useful to show also just the:
* *Actual balance*: balance